### PR TITLE
Update bundle to support Flysystem v2

### PR DESCRIPTION
**Pull request by Lustmored**
I have built upon flysystem-v2 branch and set it as a target as probably such a change will require major version bump.

Most of the work was updating tests to actually run against new architecture. I have commented and removed functionality and adapters that are missing from v2.

I haven't touched docs and didn't add support for any new options as I wanted to get feedback before eventually delving more.

This PR also includes merging master into flysystem-v2 as a base for work.

### CODE REVIEW

**Comment by tgalopin** [comment]
Thanks a lot @Lustmored !

It seems you should probably rebase on master as the first commits in your PR are newly merged commits.

I'll review this asap :) .

**Comment by Lustmored** [comment]
> Thanks a lot @Lustmored !
> 
> It seems you should probably rebase on master as the first commits in your PR are newly merged commits.
> 
> I'll review this asap :) .

Right! I have force-pushed rebase onto master.

**Review by JeroenMoonen — APPROVED** [review]

**Comment by tgalopin** [comment]
There still seems to be a merge problem, are you sure you rebased properly?

**Comment by Lustmored** [comment]
I am pretty sure. Yet as I've stated - this PR is based on 'flysystem-v2' branch from original repository and targets it. I believe rebase of 'flysystem-v2' branch onto master will fix the conflicts. Alternatively I can target 'master' branch.

**Comment by tgalopin** [comment]
Ah right! My bad, I must have badly rebased myself then :) I'll have a look asap (I'm preparing the SymfonyWorld conference right now, kinda busy but I'll have more time later :) )

**Comment by Lustmored** [comment]
I have changed PR target to `master` for easier review and merge.

I have also updated docs and re-added cached storage driver based on `lustmored/flysystem-v2-simple-cache-adapter` for easier upgrade path with useful, yet lighter, cached filesystem driver. If you'd prefer I don't just let me know and I will drop that part from the PR :)

**Comment by tgalopin** [comment]
Thanks for the ping, I'll have a look :)

**Comment by Lustmored** [comment]
@tgalopin Just a friendly remainder, no pressure intended :)

I've found out my commits authoring was off, so fixed it and force pushed. No changes in code this time.

File: composer.json [hunk 1]
```diff
@@ -20,36 +20,24 @@
         }
     },
     "require": {
-        "php": ">=7.1",
-        "league/flysystem": "^1.0.40",
+        "php": "^7.2 || ^8.0",
```
Comment (tgalopin) [inline composer.json:23]: Please keep the `>=` notation, it's easier to maintain.

File: composer.json [hunk 2]
```diff
@@ -20,36 +20,24 @@
...
-        "league/flysystem-cached-adapter": "^1.0.9",
-        "league/flysystem-memory": "^1.0",
-        "league/flysystem-rackspace": "^1.0",
-        "league/flysystem-replicate-adapter": "^1.0",
-        "league/flysystem-sftp": "^1.0",
-        "league/flysystem-webdav": "^1.0",
-        "league/flysystem-ziparchive": "^1.0",
+        "league/flysystem-aws-s3-v3": "^2.0",
+        "league/flysystem-memory": "^2.0",
+        "lustmored/flysystem-v2-simple-cache-adapter": "^0.1.0",
```
Comment (tgalopin) [inline composer.json:34]: There is no cache adapter anymore? Perhaps would it make sense to remove the adapter then?

File: docs/2-cloud-storage-providers.md [hunk 3]
```diff
@@ -4,37 +4,10 @@ One of the core feature of Flysystem is its ability to interact easily with remo
 including many cloud storage providers. This bundle provides the same level of support for these
 cloud providers by providing corresponding adapters in the configuration.
 
-* [Azure](#azure)
 * [AsyncAws S3](#asyncaws-s3)
 * [AWS S3](#aws-s3)
 * [DigitalOcean Spaces](#digitalocean-spaces)
 * [Scaleway Object Storage](#scaleway-object-storage)
-* [Google Cloud Storage](#google-cloud-storage)
```
Comment (tgalopin) [inline docs/2-cloud-storage-providers.md:12]: We are removing quite useful adapters here (Azure and Google Cloud Storage), are there community alternatives?

File: docs/4-caching-metadata-in-symfony-cache.md [hunk 4]
```diff
@@ -10,19 +10,23 @@ involves an additional overhead on top of the classical slowness of I/O.
 When your application needs to scale, you may need to cache metadata in order to
 improve performances on this level. To do this, you can use the `cache` adapter.
 
-> *Note:* this adapter caches anything but the file content. This keeps the cache 
+> *Note:* this adapter caches just the metadata of existing files. This keeps the cache
```
Comment (tgalopin) [inline docs/4-caching-metadata-in-symfony-cache.md:13]: this adapter only caches the metadata of existing files, not their content.

File: docs/4-caching-metadata-in-symfony-cache.md [hunk 5]
```diff
@@ -10,19 +10,23 @@ involves an additional overhead on top of the classical slowness of I/O.
 When your application needs to scale, you may need to cache metadata in order to
 improve performances on this level. To do this, you can use the `cache` adapter.
 
-> *Note:* this adapter caches anything but the file content. This keeps the cache 
+> *Note:* this adapter caches just the metadata of existing files. This keeps the cache 
 > small enough to be beneficial and covers all the file system inspection operations.
 
+> *Note:* Earlier versions used different cache driver that was memory heave on systems
```
Comment (tgalopin) [inline docs/4-caching-metadata-in-symfony-cache.md:16]: Unnecessary note IMO

File: docs/4-caching-metadata-in-symfony-cache.md [hunk 6]
```diff
@@ -10,19 +10,23 @@ involves an additional overhead on top of the classical slowness of I/O.
...
 ```
-composer require league/flysystem-cached-adapter
+composer require lustmored/flysystem-v2-simple-cache-adapter
 ```
 
 ### Usage
 
 The cache adapter works using a source storage (from which it will read the uncached data)
-and a [Symfony cache pool](https://symfony.com/doc/current/reference/configuration/framework.html#pools). 
+and any [PSR-6 cache item pool interface](https://www.php-fig.org/psr/psr-6/).
```
Comment (tgalopin) [inline docs/4-caching-metadata-in-symfony-cache.md:29]: A Symfony pool is a PSR-6 pool. We're making a Symfony bundle, let's keep the Symfony reference.

File: tests/Adapter/Builder/AsyncAwsAdapterDefinitionBuilderTest.php [hunk 7]
```diff
@@ -39,39 +39,39 @@ public function provideValidOptions()
...
         ]];
 
-        yield 'options' => [[
-            'client' => 'my_client',
-            'bucket' => 'bucket',
-            'options' => [
-                'ServerSideEncryption' => 'AES256',
-            ],
-        ]];
+//        yield 'options' => [[
```
Comment (tgalopin) [inline tests/Adapter/Builder/AsyncAwsAdapterDefinitionBuilderTest.php:42]: Either we should remove this code or enable it back

File: tests/Adapter/Builder/AsyncAwsAdapterDefinitionBuilderTest.php [hunk 8]
```diff
@@ -39,39 +39,39 @@ public function provideValidOptions()
...
-        ]);
-
-        $this->assertSame(S3FilesystemV1::class, $definition->getClass());
-        $this->assertInstanceOf(Reference::class, $definition->getArgument(0));
-        $this->assertSame('my_client', (string) $definition->getArgument(0));
-        $this->assertSame('bucket', $definition->getArgument(1));
-        $this->assertSame('prefix/path', $definition->getArgument(2));
-        $this->assertSame(['ServerSideEncryption' => 'AES256'], $definition->getArgument(3));
-    }
+//    public function testOptionsBehavior()
```
Comment (tgalopin) [inline tests/Adapter/Builder/AsyncAwsAdapterDefinitionBuilderTest.php:59]: Same

File: tests/Adapter/options.yaml [hunk 9]
```diff
@@ -13,24 +13,24 @@ fs_aws:
         bucket: 'bucket_name'
         prefix: 'optional/path/prefix'
 
-fs_azure:
-    adapter: 'azure'
-    options:
-        client: 'azure_client_service'
-        container: 'container_name'
-        prefix: 'optional/path/prefix'
+#fs_azure:
```
Comment (tgalopin) [inline tests/Adapter/options.yaml:16]: Same and for the next commented code too

File: tests/Adapter/Builder/LocalAdapterDefinitionBuilderTest.php [hunk 10]
```diff
@@ -78,18 +78,18 @@ public function provideValidOptions()
...
+        $this->assertSame(LocalFilesystemAdapter::class, $this->createBuilder()->createDefinition($options)->getClass());
     }
 
     public function testOptionsBehavior()
     {
         $permissions = [
             'file' => [
-                'public' => 0755,
-                'private' => 0755,
+                'public' => 0666,
```
Comment (tgalopin) [inline tests/Adapter/Builder/LocalAdapterDefinitionBuilderTest.php:88]: This feels like a bug as the config didn't change

**Review by tgalopin — COMMENTED** [review]
Hi!

Thanks for the friendly ping :) . Here is a review. Also, you should rebase your branch on master to get the latest CI fixes.

More generally: I would find it useful to keep Azure and Google Cloud support. Are there credible community alternatives for Flysytem 2?

File: docs/2-cloud-storage-providers.md (hunk 3 above)
Comment (Lustmored) [reply docs/2-cloud-storage-providers.md:12]: All Azure adapters doesn't even have issues for flysystem v2.

I just found that league published google cloud adapter recently (https://github.com/thephpleague/flysystem-google-cloud-storage) and I will work towards this adapter soon and will update PR.

File: docs/4-caching-metadata-in-symfony-cache.md (hunk 5 above)
Comment (Lustmored) [reply docs/4-caching-metadata-in-symfony-cache.md]: OK, removed

File: tests/Adapter/Builder/LocalAdapterDefinitionBuilderTest.php (hunk 10 above)
Comment (Lustmored) [reply tests/Adapter/Builder/LocalAdapterDefinitionBuilderTest.php]: Right, I was checking something during development and missed reverting this change.

File: composer.json (hunk 2 above)
Comment (Lustmored) [reply composer.json:36]: There is no upstream adapter for v2 and there seems to be no work towards it. V1 adapter also was very memory inefficient (https://github.com/thephpleague/flysystem-cached-adapter/issues/23). That is why I have created this simple alternative for v2, which is very helpful in my projects already, so I think it can be helpful for others too.

But of course if you feel like dropping it from the bundle, I can do so.

**Comment by Lustmored** [comment]
I have updated PR with Google Cloud support, rebase to master and changes requested in comments

File: composer.json (hunk 2 above)
Comment (tgalopin) [reply composer.json:37]: Let's drop it for now, we'll add it if someone find it useful (not sure it is)

File: docs/2-cloud-storage-providers.md (hunk 3 above)
Comment (tgalopin) [reply docs/2-cloud-storage-providers.md:12]: Let's implement the official Google Cloud Storage package yes, and let's ignore Azure for now: if people need it, it'll be added back.

Comment (Lustmored) [reply docs/2-cloud-storage-providers.md:12]: OK, I have already done that :)

File: composer.json (hunk 2 above)
Comment (Lustmored) [reply composer.json]: OK, I have removed it from code and docs

**Comment by tgalopin** [comment]
Thanks so much @Lustmored ! I'll have a look later today but thanks for your work :) !

**Comment by nCrazed** [comment]
@Lustmored, @tgalopin is there anything that you're missing, could use some help with here?

I was planning to add flysystem to an existing Symfony project, but decided to postpone until this bundle starts can support V2. I am happy to put in some work in order to speed it up

**Comment by Lustmored** [comment]
> @Lustmored, @tgalopin is there anything that you're missing, could use some help with here?
> 
> I was planning to add flysystem to an existing Symfony project, but decided to postpone until this bundle starts can support V2. I am happy to put in some work in order to speed it up

We have to wait until @tgalopin has time to take another look and possibly merge and release

**Comment by tgalopin** [comment]
Thanks so much for your work @Lustmored ! I'll release 2.0 soon.

File: tests/DependencyInjection/FlysystemExtensionTest.php [hunk 11]
```diff
@@ -31,19 +27,10 @@ public function provideFilesystems()
     {
         $fsNames = [
             'fs_aws',
-            'fs_azure',
-            'fs_cache',
-            'fs_custom',
-            'fs_dropbox',
             'fs_ftp',
             'fs_gcloud',
-            'fs_lazy',
```
Comment (tgalopin) [inline tests/DependencyInjection/FlysystemExtensionTest.php:40]: I missed this, why did you remove this?

File: tests/DependencyInjection/FlysystemExtensionTest.php [hunk 12]
```diff
@@ -31,19 +27,10 @@ public function provideFilesystems()
     {
         $fsNames = [
             'fs_aws',
-            'fs_azure',
-            'fs_cache',
-            'fs_custom',
```
Comment (tgalopin) [inline tests/DependencyInjection/FlysystemExtensionTest.php:36]: I missed this, why did you remove this?

File: tests/Adapter/Builder/GcloudAdapterDefinitionBuilderTest.php [hunk 13]
```diff
@@ -31,23 +31,10 @@ public function provideValidOptions()
             'bucket' => 'bucket',
         ]];
 
-        yield 'prefix' => [[
-            'client' => 'my_client',
-            'bucket' => 'bucket',
-            'prefix' => 'prefix/path',
-        ]];
-
-        yield 'api_url' => [[
```
Comment (tgalopin) [inline tests/Adapter/Builder/GcloudAdapterDefinitionBuilderTest.php:40]: I missed this, why did you remove this?

File: tests/Adapter/Builder/GcloudAdapterDefinitionBuilderTest.php [hunk 14]
```diff
@@ -31,23 +31,10 @@ public function provideValidOptions()
             'bucket' => 'bucket',
         ]];
 
-        yield 'prefix' => [[
```
Comment (tgalopin) [inline tests/Adapter/Builder/GcloudAdapterDefinitionBuilderTest.php:34]: I missed this, why did you remove this?

File: tests/Kernel/config/flysystem.yaml [hunk 15]
```diff
@@ -7,35 +7,13 @@ flysystem:
...
-                container: 'container_name'
-                prefix: 'optional/path/prefix'
-
-        fs_cache:
-            adapter: 'cache'
-            options:
-                store: 'cache.app'
-                source: 'fs_local'
-
-        fs_custom:
```
Comment (tgalopin) [inline tests/Kernel/config/flysystem.yaml:23]: I think we should keep this shouldn't we?

**Review by tgalopin — COMMENTED** [review]
A bit late, I agree, but I just realized a few things

**Comment by tgalopin** [comment]
Released

@nCrazed could you test to ensure everything is good on your side :) ? I tested on my projects but more tests can't hurt.

**Comment by nCrazed** [comment]
I am planning to start the refactoring/integration next week. Will report any problems that I run into.

**Comment by Lustmored** [comment]
@tgalopin @nCrazed if any issues arise I am more than willing to help tackling them. Please mention me in any issue related to this and I will gladly help.

**Comment by tgalopin** [comment]
Thanks @Lustmored, really appreciate the help!

**Comment by nspyke** [comment]
Hi @tgalopin 
I'm currently using this bundle and the AWS adapter. I need to move my app to Azure. Would it be possible to reinstate Azure Blob storage support?
I'd be happy to submit a PR if it were to be reviewed and merged,
